theano/gpuarray/opt.py

```diff
@@ -837,71 +1165,79 @@ def local_gpuagemmbatch_output_merge(node, *inputs):
 
 
 @register_opt('fast_compile')
 @op_lifter([tensor.blas.Dot22])
-def local_gpua_dot22(node, context_name):
+@register_opt2([tensor.blas.Dot22], 'fast_compile')
+def local_gpua_dot22(op, context_name, inputs, outputs):
     return gpu_dot22
 
 
 @register_opt('fast_compile')
 @op_lifter([tensor.blas.Dot22Scalar])
-def local_gpua_dot22scalar(node, context_name):
-    x, y, a = node.inputs
+@register_opt2([tensor.blas.Dot22Scalar], 'fast_compile')
+def local_gpua_dot22scalar(op, context_name, inputs, outputs):
+    x, y, a = inputs
     x = as_gpuarray_variable(x, context_name)
     y = as_gpuarray_variable(y, context_name)
-    z = GpuAllocEmpty(x.dtype, context_name)(x.shape[0], y.shape[1])
+    z = gpu_alloc_empty(x.dtype, context_name)(x.shape[0], y.shape[1])
     return [gpugemm_no_inplace(z, a, x, y, 0)]
 
 
 @register_opt('fast_compile')
 @op_lifter([tensor.basic.Eye])
-def local_gpua_eye(node, context_name):
-    return GpuEye(dtype=node.op.dtype, context_name=context_name)
+@register_opt2([tensor.basic.Eye], 'fast_compile')
+def local_gpua_eye(op, context_name, inputs, outputs):
+    return GpuEye(dtype=op.dtype, context_name=context_name)
 
 
 @register_opt('fast_compile')
 @op_lifter([tensor.nnet.CrossentropySoftmaxArgmax1HotWithBias], cuda_only=True)
-def local_gpua_crossentropysoftmaxargmax1hotwithbias(node, context_name):
+@register_opt2([tensor.nnet.CrossentropySoftmaxArgmax1HotWithBias], 'fast_compile')
+def local_gpua_crossentropysoftmaxargmax1hotwithbias(op, context_name, inputs, outputs):
     return gpu_crossentropy_softmax_argmax_1hot_with_bias
 
 
 @register_opt('fast_compile')
 @op_lifter([tensor.nnet.CrossentropySoftmax1HotWithBiasDx], cuda_only=True)
-def local_gpua_crossentropysoftmax1hotwithbiasdx(node, context_name):
+@register_opt2([tensor.nnet.CrossentropySoftmax1HotWithBiasDx], 'fast_compile')
+def local_gpua_crossentropysoftmax1hotwithbiasdx(op, context_name, inputs, outputs):
     return gpu_crossentropy_softmax_1hot_with_bias_dx
 
 
 @register_opt('fast_compile')
 @op_lifter([tensor.nnet.Softmax], cuda_only=True)
-def local_gpua_softmax(node, context_name):
+@register_opt2([tensor.nnet.Softmax], 'fast_compile')
+def local_gpua_softmax(op, context_name, inputs, outputs):
     return gpu_softmax
 
 
 @register_opt('fast_compile')
 @op_lifter([tensor.nnet.SoftmaxWithBias], cuda_only=True)
-def local_gpua_softmaxwithbias(node, context_name):
+@register_opt2([tensor.nnet.SoftmaxWithBias], 'fast_compile')
+def local_gpua_softmaxwithbias(op, context_name, inputs, outputs):
     return gpu_softmax_with_bias
 
 
 @register_opt('fast_compile')
 @op_lifter([theano.tensor.opt.Assert])
-def local_assert(node, context_name):
-
+@register_opt2([theano.tensor.opt.Assert], 'fast_compile')
+def local_assert(op, context_name, inputs, outputs):
     # Check if input nodes are already on the GPU
-    if isinstance(node.inputs[0].type, GpuArrayType):
+    if isinstance(inputs[0].type, GpuArrayType):
         return
```

Comment:
I think we should do the same here: split the function, remove the check for GpuArrayType from the function decorated with `@register_opt2`, add it to the one decorated by `@op_lifter`.